File: lib/services/mail.ts
import { Resend } from "resend";

const senderEmail = process.env.SENDER_EMAIL || "no-reply@solvex-eui.org";
const senderName = "Solve-The-X";

// Initialize Resend client
const resendApiKey = process.env.RESEND_API_KEY || "";
const resend = new Resend(resendApiKey);

/**
 * Send email via Resend with diagnostic logging
 */
async function sendViaResend(to: string, subject: string, text: string, html: string) {
  console.log("📧 [Resend] Attempting to send email:", {
    from: `${senderName} <${senderEmail}>`,
    to,
    subject,
  });

  try {
    const result = await resend.emails.send({
      from: `${senderName} <${senderEmail}>`,
      to: Array.isArray(to) ? to : [to],
      subject,
      html,
      text,
    });
    
    console.log("✅ [Resend] Email sent successfully:", result);
    return result;
  } catch (error) {
    console.error("❌ [Resend] Email send failed:", error);
    throw error;
  }
}

/**
 * Send verification code email
 */
export async function sendVerificationEmail(to: string, code: string) {
  const subject = "Verify your account";
  const text = `Your verification code is: ${code}`;
  const html = `<p>Your verification code is: <b>${code}</b></p>`;

  console.log("📨 [Mail Service] Sending verification email with code:", code.substring(0, 2) + "****");
  return sendViaResend(to, subject, text, html);
}

/**
 * Send notification email
 */
export async function sendNotificationEmail(
  to: string,
  subject: string,
  content: string
) {
  const html = `<p>${content}</p>`;

  console.log("📨 [Mail Service] Sending notification email:", subject);
  return sendViaResend(to, subject, content, html);
}

File: solve-x-main/backend/mail_service/mail.go
package mail_service

import (
	"fmt"
	"os"
	"strconv"

	"github.com/sendgrid/sendgrid-go"
	"github.com/sendgrid/sendgrid-go/helpers/mail"
	"gopkg.in/gomail.v2"
)

// Mailman handles sending emails via SendGrid or SMTP
type Mailman struct {
	useSendGrid bool
	senderEmail string
	senderName  string
	// SendGrid fields
	sendGridAPIKey string
	// SMTP fields
	smtpHost     string
	smtpPort     string
	smtpUsername string
	smtpPassword string
}

// NewMailman creates a new Mailman instance based on MAIL_SERVICE env variable
func NewMailman() *Mailman {
	mailService := os.Getenv("MAIL_SERVICE")
	useSendGrid := mailService != "local"

	return &Mailman{
		useSendGrid:    useSendGrid,
		senderEmail:    os.Getenv("SENDER_EMAIL"),
		senderName:     "Solve-The-X",
		sendGridAPIKey: os.Getenv("SENDGRID_API_KEY"),
		smtpHost:       os.Getenv("SMTP_HOST"),
		smtpPort:       os.Getenv("SMTP_PORT"),
		smtpUsername:   os.Getenv("SMTP_USERNAME"),
		smtpPassword:   os.Getenv("SMTP_PASSWORD"),
	}
}

// SendVerificationEmail sends a verification code email
func (m *Mailman) SendVerificationEmail(to, code string) error {
	subject := "Verify your account"
	plainText := fmt.Sprintf("Your verification code is: %s", code)
	htmlContent := fmt.Sprintf("<p>Your verification code is: <b>%s</b></p>", code)

	if m.useSendGrid {
		return m.sendViaSendGrid(to, subject, plainText, htmlContent)
	}
	return m.sendViaSMTP(to, subject, plainText, htmlContent)
}

// SendNotification sends a notification email
func (m *Mailman) SendNotification(to, subject, content string) error {
	htmlContent := "<p>" + content + "</p>"

	if m.useSendGrid {
		return m.sendViaSendGrid(to, subject, content, htmlContent)
	}
	return m.sendViaSMTP(to, subject, content, htmlContent)
}

// sendViaSendGrid sends email using SendGrid API
func (m *Mailman) sendViaSendGrid(to, subject, plainText, htmlContent string) error {
	from := mail.NewEmail(m.senderName, m.senderEmail)
	toEmail := mail.NewEmail("", to)

	message := mail.NewSingleEmail(from, subject, toEmail, plainText, htmlContent)
	client := sendgrid.NewSendClient(m.sendGridAPIKey)
	response, err := client.Send(message)
	if err != nil {
		return fmt.Errorf("sendgrid error: %w", err)
	}

	if response.StatusCode >= 400 {
		return fmt.Errorf("sendgrid error: %d - %s", response.StatusCode, response.Body)
	}

	return nil
}

// sendViaSMTP sends email using gomail package
func (m *Mailman) sendViaSMTP(to, subject, plainText, htmlContent string) error {
	// Convert port to int
	port, err := strconv.Atoi(m.smtpPort)
	if err != nil {
		return fmt.Errorf("invalid SMTP port: %w", err)
	}

	// Create new message
	msg := gomail.NewMessage()
	msg.SetHeader("From", fmt.Sprintf("%s <%s>", m.senderName, m.senderEmail))
	msg.SetHeader("To", to)
	msg.SetHeader("Subject", subject)
	msg.SetBody("text/plain", plainText)
	msg.AddAlternative("text/html", htmlContent)

	// Create dialer
	dialer := gomail.NewDialer(m.smtpHost, port, m.smtpUsername, m.smtpPassword)

	// For local development without auth
	if m.smtpUsername == "" && m.smtpPassword == "" {
		dialer.Auth = nil
	}

	// Send email
	if err := dialer.DialAndSend(msg); err != nil {
		return fmt.Errorf("smtp error: %w", err)
	}

	return nil
}

// Legacy functions for backward compatibility
func SendVerificationEmail(to, code string) error {
	mailman := NewMailman()
	return mailman.SendVerificationEmail(to, code)
}

func SendNotification(to, subject, content string) error {
	mailman := NewMailman()
	return mailman.SendNotification(to, subject, content)
}
